### PR TITLE
fix(rust-core): close all Python/Rust parity gaps — SILICON audit + HYBRID/EMPIRICAL port

### PR DESCRIPTION
**Pull request by tianhaox**
## Summary

Full-surface parity between the Python SDK and `rust/aiconfigurator-core`, in two halves merged into one branch:

1. **SILICON-path audit**: closes every divergence found by a fresh whole-repo audit (audit doc + per-item status: `rust/aiconfigurator-core/docs/parity-audit-2026-07-14.md`).
2. **HYBRID/EMPIRICAL (`util_empirical`) port** (issue #1333 §4.6 option b): the compiled engine now implements the full util-space empirical layer — `latency = SOL(query)/util` with the xshape/xquant/xprofile/xop transfer ladder, transfer-policy gating, and `EmpiricalNotImplemented` honest-miss semantics — so `--database-mode HYBRID/EMPIRICAL` runs on the Rust step instead of delegating to Python. Only the SOL/SOL_FULL diagnostic modes still delegate.

Out of scope remain the by-design Python-only surfaces (AFD, vision-encoder phase, power/energy, `.txt` legacy loading, scheduling/summary orchestration).

Every fix lands with a pinning test, and each previously-uncovered trigger config is now in the parity matrices.

## P0 fixes (live numeric divergences)

| Area | Bug | Fix |
|---|---|---|
| WideEP MLA | opspec emitted raw `tp_size` into the per-rank-heads table axis (tp=8 queried heads=8 instead of 16 → off-grid extrapolation) | emit `128 // tp_size`; mirror the `{flashinfer, fa3}` whitelist |
| wideep loaders | Rust first-wins vs Python last-wins dedup (270 real dup keys on rtx_pro_6000_server) | last-wins across wideep_moe + wideep context/generation MoE |
| DSv4 | #1337 was ported to `mla.rs` but not `dsv4.rs`; oracles certified stale behavior | generation table keys `[kv][gemm]` only, SOL fmha derived from kv; arch key dropped; oracles regenerated |
| trtllm alltoall | shared loader collapsed 1,556/2,096 gb200 rows (dropped kernel_source/op_name/num_nodes) and queried a `distribution` axis Python ignores → hard error on shipped data | dedicated loader with Python's exact keying; `_select_alltoall_kernel` port; fp8_block→fp8; full trtllm dispatch branch (sm==100 gate — gb300 is sm103, dp>1 volumes, NVL72 reroute) |
| mixed/decode step | three composition gaps (pass-2 chunked-prefill query shape, pass-1 floor/nextn accounting, ±1 kv convention) | rewritten as a literal mirror of Python's three-pass `_get_mix_step_latency` — **bit-identical** on chunked/prefix/MTP shapes |
| imbalance scales | accepted on the wire, hardcoded 1.0 in every `RuntimeContext` | threaded FFI → session → attention ops |
| DSA (GLM-5.2) | skip-indexer rows dropped at load; no `full_frac` amortization → silent overestimate | skip tables + skip SOL terms + CP semantics + `full_frac` blend |
| sglang MoE | deepep compute read `moe_perf`; EPLB ×0.8 dropped; mxfp4 remaps missing; DeepEP dispatch flavors never emitted | all four routed/ported; `scale_num_tokens` on the wire |
| TrtLLMWideEPMoEDispatch | no opspec branch → trtllm WideEP DeepSeek failed conversion outright | new `WideEpMoeDispatch` op (prepare+dispatch / combine) |

Plus the P1 batch: GEMM fp8_static SOL floor, elementwise integer ordering, fractional NCCL message sizes, encoder partial-RoPE, wideep num_slots roofline holds, beam-width token scaling, NVL72 reroute sunk to the DB layer.

## Robustness / infra

- `OpConversionError` falls back to the Python step (parity by delegation, memoized) instead of crashing the sweep.
- Engine-handle cache identity no longer aliases collapsed dtypes (`sq` vs `int8_wo`, …) or identity-omitted ModelConfig fields; the key is memoized on the model object (the per-step `json.dumps` cost tripped the perf gate mid-branch and was fixed).
- `cargo test --lib` joins the parity CI job (196 tests): two stale pins fixed (b200 mem_bw 7.7 TB/s survived PR #1246 unnoticed), py:: binding tests self-initialize the interpreter.
- Enum mirrors corrected (ModelFamily GEMMA4MIX/MINIMAXM3, CommQuantMode int8/fp8, MoEQuantMode mxfp4 members, PerfDataFilename → parquet); wrong-composition caller-less `overlap_composition` deleted; shared-layer resolution failures now warn loudly.

## New parity coverage

- GLM-5.2 skip-indexer smoke case (SMOKE_CASES)
- sglang WideEP/DeepEP end-to-end suite (h200/0.5.6.post2) — bit-identical
- trtllm WideEP alltoall suite (gb200/1.3.0rc10) — bit-identical
- chunked-prefill mixed-step cases — bit-identical
- non-1.0 imbalance-scale threading cases

## Validation

| Suite | Result |
|---|---|
| engine-step parity | 235/235 |
| compile-engine parity | 55/55 |
| perf regression gate | 3/3 |
| cargo test --lib | 196/196 |
| sdk unit tests | 1323/1323 |

## Follow-ups (not in this PR)

- Full-matrix `scan_rust_parity.py` run before merge (hours; not run locally).
- Gate 2 golden-oracle capture (python-dedup-plan.md) before any Python deletion.
- Default flip (Gate 1) stays blocked on the #1333 hybrid port.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added imbalance-correction scale parameters to mixed-step and decode-step latency APIs and introduced provenance reporting.
  * Expanded performance modeling with database-mode + transfer-policy routing, plus broader op coverage (e.g., MSA, WideEP MoE/Alltoall, DSA skip-indexer, Partial-RoPE).

* **Bug Fixes**
  * Rust engine-step compilation gaps now fall back to Python estimation instead of failing.
  * Hybrid/empirical handling now surfaces explicit “not implemented” coverage gaps and improves dispatch consistency.

* **Tests**
  * Expanded Rust↔Python parity and added regression/oracle coverage for new shapes, scales, and HYBRID/EMPIRICAL edge cases.

* **Chores**
  * CI now runs Rust unit tests during the parity workflow.

## HYBRID/EMPIRICAL port (merged from the hybrid-rust-parity branch)

Test-first: 16 new `HYBRID_CASES` in `test_engine_step_parity.py` (×4 surfaces = 64 checks) were landed as a red baseline (46 failing) before the port, each tier pinned by a `capture_provenance()`-verified real-data config:

| tier | case |
|---|---|
| xop + xshape | MiniMax-M3 (MSA ← DSA util borrow) on vllm 0.19.0 AND sglang 0.5.14 |
| xquant | Qwen3-235B + forced `moe_quant=w4a16_mxfp4_cutlass` (same (0.5,1) profile as collected int4_wo) |
| xprofile | Qwen3-235B + forced `w4afp8` (no same-profile sibling; util-level rescale) |
| attention xshape | MiMo-V2-Flash (head_dim=192; collected {128,256}) |
| own-data empirical | EMPIRICAL mode across 7 op families (dense/MoE/MLA/DSA×2/fp8_block/state-space) |
| contracts | HYBRID==SILICON invariance (Kimi), policy off/balanced → symmetric `EmpiricalNotImplementedError`, ladder miss (NVFP4 on Hopper) |

Hybrid cases assert at **rtol=1e-4** (not the 1% smoke tolerance): EMPIRICAL≈SILICON by design on collected shapes — and identical at exact grid hits — so 1% cannot distinguish a mode-ignoring engine.

Mechanics:
- `operators/util_empirical.rs`: `UtilGrid` (k=2 IDW in per-axis normalised log space, exact-hit preservation, boundary clamp), `ZeroAwareDeltaLookup` (compute_scale deltas), keyed grid caches. Unit tests mirror `tests/unit/sdk/test_util_empirical.py`.
- Per-family mode dispatch (EMPIRICAL → estimate; HYBRID → silicon, typed miss → estimate; `EmpiricalNotImplemented` never caught) for GEMM/attention/MoE-ladder/MLA/DSA/DSV4/MHC/communication/alltoall/wideep-MoE, each anchored against Python-generated oracles at 1e-9 in `#[cfg(test)]`.
- `PerfDatabase` is now a mode-configured view over `Arc<PerfTables>` (Deref keeps table access unchanged); `silicon_view()` gives `Op::Fallback` Python's semantics (under HYBRID the primary is evaluated silicon-only) and powers MSA's cross-op DSA probe.
- New `Op::MsaContext/MsaGeneration` + opspec conversion — MiniMax-M3 compiles and matches at 1e-4 end-to-end. `ENGINE_SPEC_SCHEMA_VERSION` → 3 (variant insertion shifted bincode indices).
- `EngineConfig` carries `database_mode` + `transfer_policy` (serde defaults keep old specs valid); the handle cache key includes both.
- mamba/GDN verified mode-independent in Python (SOL-degradation contract) — no change needed. MegaMoE has no Python empirical path — none ported.

Integration notes (merge of the two branches):
- `perf_database/wideep_moe` keeps the hybrid branch's level-wise distribution resolution (kernel → quant → distribution, then exact shape) — the any-shape fallback was verified divergent against Python.
- `WideEpMoeOp` re-resolves the compute kernel at **query time** (Python `_select_kernel(database, quant)` needs loaded tables; the compile-time opspec value is load-order dependent and broke the TrtllmWideEp compile-parity case on a fresh database). The opspec `kernel_source` stays as a wire hint.

## Post-merge audit (capability + selection-rule) — findings closed

A whole-surface audit after the merge found and fixed:
- **6 selection-rule divergences** (2 live on shipped data: the WideEP MoE distribution-fallback file-order pick; the MoE xprofile fp8/fp8_block tie on vllm 0.24.0 — a measured ~13% divergence), plus deepep EMPIRICAL calibration routing, EPLB correction scope, the WideEP-MLA whitelist placement, and the GB200/NVL72 reroute mode-awareness.
- **Typed errors across the FFI**: rust misses now raise the sdk `PerfDataNotAvailableError` / `EmpiricalNotImplementedError` (classifier tools work on the rust path).
- **Empirical provenance across the FFI**: the transfer tier (empirical/xshape/xquant/xprofile/xop) is recorded rust-side and fed into `capture_provenance`, so support-matrix `HYBRID_PASS` labelling works on rust-routed runs.
- **DSV4 MegaMoE op ported** (tail-appended variant, no schema bump) — megamoe models no longer fall back to the Python step.
- **Python fix**: the trtllm-alltoall HYBRID fallback closure passed 7/8 args to `get_empirical_from_sol` (latent `TypeError` on every HYBRID silicon miss) + regression test.

Deliberately deferred: energy/power over the FFI (rust-routed reports show `power=0.0W`; follow-up PR — threads the energy column through the interp core, every loader, every op). Rust stays parquet-only.

## Verification (final tree)

- `rust/aiconfigurator-core/parity_tests/`: **361 passed** (engine-step 303 — 64 hybrid surfaces at rtol=1e-4, full SILICON suite, new typed-error + provenance-capture tests; compile-engine parity; perf gate)
- `cargo test --lib`: **276 passed, 0 failed**
- Python unit suite: **2077 passed**

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: rust/aiconfigurator-core/src/common/enums.rs [hunk 1]
```diff
@@ -66,6 +66,78 @@ impl Default for DatabaseMode {
...
+/// (`off`/`conservative`/`balanced`/`aggressive`) before serialising, so the
+/// wire form is always an explicit list of kind tokens; `None` on the wire
+/// means the default ALL-transfers policy.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Default)]
+pub struct TransferPolicy {
+    pub xshape: bool,
+    pub xquant: bool,
+    pub xprofile: bool,
+    pub xop: bool,
+}
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply rust/aiconfigurator-core/src/common/enums.rs]: Fixed in bef449250 — `TransferPolicy` now has an explicit `impl Default` returning `Self::ALL`, matching the documented unset-means-ALL wire semantics.

File: aic-core/rust/aiconfigurator-core/src/common/error.rs [hunk 2]
```diff
@@ -54,3 +60,16 @@ pub enum AicError {
...
+    /// `_MISSING_SILICON_DATA_EXCEPTIONS` (`PerfDataNotAvailableError`,
+    /// `InterpolationDataNotAvailableError`) — the same set `Op::Fallback`
+    /// catches. `EmpiricalNotImplemented` is deliberately NOT in this set:
+    /// it is the terminal miss raised after the empirical path itself found
+    /// no calibration data.
+    pub fn is_missing_perf_data(&self) -> bool {
+        matches!(self, Self::PerfDatabase(_) | Self::Io { .. })
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply aic-core/rust/aiconfigurator-core/src/common/error.rs:75]: Agreed directionally — Python only treats `PerfDataNotAvailableError`/`InterpolationDataNotAvailableError` as fallback-able (`_MISSING_SILICON_DATA_EXCEPTIONS`, perf_database.py:29); OSErrors surface as hard errors. But the one-line narrowing is ineffective on its own: every table loader memoizes load failures behind `clone_err`, which collapses ALL variants (incl. `Parquet`/`Io`) to `PerfDatabase(msg)` before `is_missing_perf_data` ever sees them (13 identical `clone_err` sites), and several loaders match `Io | PerfDatabase` directly (gemm.rs:769, mhc.rs:304, mla.rs:1020, dsv4.rs:1301, wideep.rs:1459, op.rs:374). The proper fix is an error-classification pass that preserves the class through the caches, then narrows Io to NotFound — error-path only, no latency effect on valid data. Deferred to the freeze follow-up: https://github.com/ai-dynamo/aiconfigurator/issues/1357

File: aic-core/rust/aiconfigurator-core/src/operators/msa.rs [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+                    _ => Err(AicError::EmpiricalNotImplemented(format!(
+                        "MSA context: no DSA util to transfer from (arch={}, b={batch_size}, \
+                         s={s}); collect DSA data or set msa_dsa_scale_k against an available \
+                         quant.",
+                        self.dsa_architecture
+                    ))),
+                }
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply aic-core/rust/aiconfigurator-core/src/operators/msa.rs:103]: Declining — Python is the parity spec here and it deliberately has neither shaping: `msa.py:298` and `msa.py:336` compute `lat = sol / (util * self._dsa_scale_k)` with no clamp and no positivity guard, and the SOL branch returns `sol * self._scale_factor` unclamped. Adding `.clamp_non_negative()` or a `dsa_scale_k > 0` guard would diverge from the frozen Python reference (`.claude/rules/rust-core/parity.md`). `dsa_scale_k` is a builder-set level-alignment constant (default 1.0), not user input; if validation is wanted it must land on both sides with an oracle first.

File: aic-core/rust/aiconfigurator-core/parity_tests/test_engine_step_parity.py [hunk 4]
```diff
@@ -1019,3 +1062,238 @@ def test_cp_parity(
...
+        ),
+        id="minimax-m3-b200-vllm-019-hybrid-policy-off",
+    ),
+    pytest.param(
+        EngineStepParityCase(
+            model_path="MiniMaxAI/MiniMax-M3",
+            database_mode="HYBRID",
+            transfer_policy="balanced",
+        ),
+        id="minimax-m3-b200-vllm-019-hybrid-policy-balanced",
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply aic-core/rust/aiconfigurator-core/parity_tests/test_engine_step_parity.py:1140]: Fixed in bef449250 — `_parity_mismatch_reason` now compares `_ErrorSentinel.kind` when both sides raise and fails on a mismatch, so a panic/TypeError paired with a typed miss no longer counts as symmetry. The typed-FFI mapping guarantees the sdk class names line up on both sides; the full 303-case engine-step suite is green under the stricter rule. (a950305 added the typed-class FFI tests; the shared comparator itself was still kind-agnostic until this commit.)

File: aic-core/rust/aiconfigurator-core/src/common/error.rs [hunk 5]
```diff
@@ -54,3 +60,16 @@ pub enum AicError {
...
+impl AicError {
+    /// The "missing silicon data" class that HYBRID converts into an
+    /// empirical estimate. Mirrors Python's
+    /// `_MISSING_SILICON_DATA_EXCEPTIONS` (`PerfDataNotAvailableError`,
+    /// `InterpolationDataNotAvailableError`) — the same set `Op::Fallback`
+    /// catches. `EmpiricalNotImplemented` is deliberately NOT in this set:
+    /// it is the terminal miss raised after the empirical path itself found
+    /// no calibration data.
+    pub fn is_missing_perf_data(&self) -> bool {
+        matches!(self, Self::PerfDatabase(_) | Self::Io { .. })
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply aic-core/rust/aiconfigurator-core/src/common/error.rs:72]: Same finding as the sibling comment on lines 63-75 — deferred to https://github.com/ai-dynamo/aiconfigurator/issues/1357: the narrowing alone is ineffective because the loaders' `clone_err` memoization collapses all variants to `PerfDatabase` before classification reaches `is_missing_perf_data`; it needs the error-classification pass described there (preserve the class through the caches, then narrow Io to NotFound).

File: aic-core/rust/aiconfigurator-core/src/perf_database/attention.rs [hunk 6]
```diff
@@ -255,6 +255,153 @@ impl AttentionTable {
...
+            }
+        }
+        if sizes.is_empty() {
+            return Err(AicError::PerfDatabase(format!(
+                "context attention data missing for fmha={fmha}, kv={kv}, \
+                 n_kv={n_kv_lookup} at {}",
+                self.data_root.display()
+            )));
+        }
+        Ok(sizes)
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply aic-core/rust/aiconfigurator-core/src/perf_database/attention.rs:321]: Declining — this mirrors the Python selection rule exactly: `_ctx_headsize_ref_grid` picks `ref_hs` from `require_data_slice(wrapper, fmha, kv, n_kv).keys()` (operations/attention.py:101), i.e. the head_size level of the nested dict aggregated across ALL windows, and only then queries the `(ref_hs, window)` slice — the exact-window -> window-0 borrow happens at the slice level (attention.py:355-361), not at candidate selection. `_gen_headsize_ref_grid` does the same (attention.py:152). Filtering candidates by window on the Rust side would change reference selection vs the frozen Python spec; if the selection rule should improve, it changes in Python first with a both-sides oracle per `.claude/rules/rust-core/parity.md`.

File: aic-core/rust/aiconfigurator-core/src/perf_database/wideep_moe.rs [hunk 7]
```diff
@@ -121,79 +120,167 @@ impl WideEpMoeTable {
...
+                 quant={quant_name:?} at {}",
                 self.data_root.display()
             )));
         }
-        query_token_curve(by_tokens, num_tokens as f64, &|t| t)
+        let dist = if requested_distribution_seen {
+            distribution
+        } else {
+            first_distribution.expect("quant_seen implies at least one distribution")
+        };
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply aic-core/rust/aiconfigurator-core/src/perf_database/wideep_moe.rs:266]: Addressed in a950305 for the load-bearing selection order: the wideep_moe kernel/distribution fallback now uses the load-time `first_distribution` record and the MoE quant enumeration uses `quants_in_load_order`, per the "load-order record" pattern in `.claude/rules/rust-core/parity.md`. The residual sorted enumerations (`sibling_slices`, `node_points` sample flattening) are observable only on exact distance ties and carry in-code NOTEs; any real tie surfacing in parity is tracked under https://github.com/ai-dynamo/aiconfigurator/issues/1357.

File: aic-core/rust/aiconfigurator-core/src/perf_database/dsv4_megamoe.rs [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+        query_token_curve(curve, f64::from(num_tokens), &|t| t)
+    }
+
+    fn load_module(&self) -> Result<&Dsv4MegaMoeGrids, AicError> {
+        let cell = self
+            .module
+            .get_or_init(|| load_module_parquet(&self.primary_path));
+        cell.as_ref().map_err(clone_err)
+    }
+}
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply aic-core/rust/aiconfigurator-core/src/perf_database/dsv4_megamoe.rs:165]: Valid — `clone_err` collapsing every cached failure to `AicError::PerfDatabase` erases the Parquet/corruption classification, and the identical pattern exists at 13 loader sites, so a local fix here would just move the inconsistency. It is the same error-classification cluster as the `is_missing_perf_data` Io-narrowing comments: preserve the class through the caches, then narrow Io to NotFound. Error-path only — no latency effect on valid data. Deferred to https://github.com/ai-dynamo/aiconfigurator/issues/1357

File: .claude/rules/rust-core/parity.md [hunk 9]
```diff
@@ -0,0 +1,59 @@
+# Rust-Core Parity Discipline (Python-path freeze)
+
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply .claude/rules/rust-core/parity.md:17]: Fixed in bef449250 — added description + `paths:` frontmatter scoping the rule to rust/aiconfigurator-core, the frozen Python op/query surfaces, and the opspec coverage tripwire.

File: tests/unit/sdk/test_opspec_coverage.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+import re
+
+import pytest
+
+pytestmark = pytest.mark.unit
+
+# Op classes deliberately NOT convertible to the compiled engine. Every entry
+# needs a reason; removing the op or porting it must also remove the entry
+# (the staleness assertion below enforces that).
+EXEMPT: dict[str, str] = {
```
[the remark this answers is not in the document]
Comment (tianhaox) [reply tests/unit/sdk/test_opspec_coverage.py:30]: Fixed in bef449250 — the test now asserts every EXEMPT entry names a live Operation class (deleted/renamed ops fail the gate) and carries a non-empty justification.

**Comment by tianhaox** [comment]
Responses to the review items that have no inline thread (outside-diff and review-body comments). Fixes are in bef449250; deferred items are tracked on #1357.

**Fixed (bef449250)**

- `rust/aiconfigurator-core/src/operators/moe_dispatch.rs:117-141` (route standalone WideEP dispatch through mode-aware querying): fixed — `TrtllmWideEpMoEDispatchOp::query` now routes through `query_alltoall_table`, matching Python `TrtLLMWideEPMoEDispatch.query` (which passes `database_mode=None`, i.e. the view's mode); phase sources combine like `PerformanceResult.__add__`. Anchored by a new Rust test (`standalone_wideep_dispatch_is_mode_aware`) composed from the existing gb200 Python oracles; SILICON values unchanged.
- `src/aiconfigurator/sdk/rust_engine_step.py:309-330` and `:333-356` (`x or 1.0` discards an explicit `0.0` scale): fixed — added `_scale_or_one` (defaults to 1.0 only for `None`) and applied it at the static, mixed-step, and decode-step wrappers.
- `rust/aiconfigurator-core/src/operators/wideep_mla.rs:318-330` (assert the whitelist-specific failure): fixed — both context and generation assertions now pin `AicError::InvalidEngineConfig` instead of `is_err()`.
- `rust/aiconfigurator-core/src/operators/dsa.rs:144-161` (stale `query_context_cp` doc): fixed — the doc now says `skip_indexer` passes through for the GLM-5.2 CP blend.

**Rejected (stale or contrary to the Python spec)**

- `rust/aiconfigurator-core/src/operators/moe.rs:247-249` (bump `ENGINE_SPEC_SCHEMA_VERSION` for new MoeOp fields): already done in this PR — the fields shipped together with the bump to 3 on both sides (`sdk/engine.py:98`, `rust/aiconfigurator-core/src/config.rs:26`); older blobs are rejected by the version gate in `engine/spec.rs`.
- `rust/aiconfigurator-core/src/operators/communication.rs:63-82` (remove unused `p2p_bandwidth`): it is live — called at lines 151 and 306.
- `rust/aiconfigurator-core/src/perf_database/communication.rs:147-199` (extract shared bandwidth-ratio helper): declining — the two functions are line-for-line mirrors of two distinct Python query paths (`CustomAllReduce._query_custom_allreduce_table`, `NCCL._query_nccl_table`), and rust-core deliberately keeps per-path verbatim mirroring so each stays diffable against its Python counterpart (parity.md). Formula changes go through Python first, per path.
- `tests/unit/sdk/test_rust_engine_step.py:377-417` (enumerate all 8 new identity fields): declining per repo testing policy (no per-field passthrough assertions); the two asserted fields were actual incident regressions, and the parity suite's cases that differ only in mode/policy would collide on identity drops of the mode fields.
- `tests/unit/sdk/test_opspec_coverage.py:65` (regex couples to one `isinstance` spelling): intentional fail-closed tripwire — `_to_opspec` branches are uniformly spelled, and a differently-spelled valid branch fails loudly and prompts an update rather than passing silently.

**Deferred to #1357**

- `rust/aiconfigurator-core/src/operators/dsa.rs:211-243` (clippy `too_many_arguments` allow): clippy is not an enforced gate on this crate (59 pre-existing warnings, incl. the same lint on `runtime.rs:372`, `moe.rs:223`, `moe_dispatch.rs:146`, `attention.rs:138`); a uniform lint-hygiene pass is the right vehicle.
- `rust/aiconfigurator-core/src/operators/util_empirical.rs:272-305, 416-441` (builders run while holding the mutex): the engine evaluates steps single-threaded today, and two-phase locking would permit duplicate expensive builds; revisit if concurrent sweeps over one shared `PerfDatabase` land.
- `.github/workflows/build-test.yml:236-245` (cargo test recompilation/caching): acknowledged; the job is within budget today — will add cargo/target caching if it trends long.
- Test-coverage suggestions: `dsv4_megamoe.rs` loader-validation synthetic tests, `runtime.rs` `beam_width > 1` case, `perf_database/moe.rs` MXFP4 remap fixture test — noted as follow-up coverage work.

File: aic-core/rust/aiconfigurator-core/docs/parity-audit-2026-07-14.md [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+> `_get_mix_step_latency` (bit-identical on chunked/prefix/MTP shapes).
+> `cargo test --lib` (196 tests, incl. regenerated oracles) is now a CI gate.
+> `OpConversionError` now falls back to the Python step instead of crashing;
+> the engine-handle cache key carries the un-collapsed quant + ModelConfig
+> identity. See the fix-order list at the bottom — items 1-12 all landed.
+
+**Scope exclusion:** HYBRID / EMPIRICAL / `util_empirical` differences are out of
+scope (Rust is SILICON-only by design; delegation gate at
+`sdk/rust_engine_step.py:224-246`; port tracked in issue #1333 with a separate
+PR in flight).
```
Comment (simone-chen) [inline aic-core/rust/aiconfigurator-core/docs/parity-audit-2026-07-14.md:24]: If the PR aims for HYBRID/EMPIRICAL port, should we include the parity scan in this PR?

File: aic-core/rust/aiconfigurator-core/parity_tests/test_engine_step_parity.py [hunk 12]
```diff
@@ -1019,3 +1077,339 @@ def test_cp_parity(
...
+        )
+        with pytest.raises(errors.EmpiricalNotImplementedError) as excinfo:
+            _rust_static_breakdown(case)
+        message = str(excinfo.value)
+        assert "empirical estimation not implemented" in message, message
+        assert not perf_database.has_perf_data_not_available_cause(excinfo.value)
+
+
+class TestRustProvenanceCapture:
+    """FFI provenance contract (audit finding #6): the compiled engine records
```
Comment (simone-chen) [inline aic-core/rust/aiconfigurator-core/parity_tests/test_engine_step_parity.py]: Should we avoid having "audit finding" in the codebase?

File: src/aiconfigurator/sdk/engine.py [hunk 13]
```diff
@@ -86,12 +91,15 @@
...
 # (`ENGINE_SPEC_SCHEMA_VERSION` / `ENGINE_CONFIG_SCHEMA_VERSION` in `lib.rs`).
 # ENGINE_SPEC bumped to 2 for the 0.10.0 op-payload layout change (CP + perf-DB
-# refactor added serialized `OpSpec` fields); the Rust consumer gates on this
-# version before decoding the positional op lists. Keep in lockstep with the
-# Rust `ENGINE_SPEC_SCHEMA_VERSION`.
-ENGINE_SPEC_SCHEMA_VERSION = 2
+# refactor added serialized `OpSpec` fields), and to 3 when the MSA op variants
+# were inserted (bincode enum indices after `DsaGeneration` shifted); the Rust
+# consumer gates on this version before decoding the positional op lists. Keep
+# in lockstep with the Rust `ENGINE_SPEC_SCHEMA_VERSION`.
```
Comment (simone-chen) [inline src/aiconfigurator/sdk/engine.py]: Shall we refactor the paragraph into bullet points and map the schema version to a specifc version+commit?

**Review by simone-chen — APPROVED** [review]

**Review by jasonqinzhou — APPROVED** [review]

File: aic-core/rust/aiconfigurator-core/docs/parity-audit-2026-07-14.md (hunk 11 above)
Comment (tianhaox) [reply aic-core/rust/aiconfigurator-core/docs/parity-audit-2026-07-14.md:24]: Not in this PR — the full-matrix scan (`tools/support_matrix/scan_rust_parity.py`) takes multiple hours per run, so it isn't practical as a PR gate. What is verified here: the 303-case engine-step parity suite (now covering HYBRID/EMPIRICAL cases and error symmetry), the full cargo suite, and the standing June full-matrix SILICON gate (2,158 entries). A full HYBRID-inclusive matrix scan is queued as a follow-up before the rust default flip.

File: aic-core/rust/aiconfigurator-core/parity_tests/test_engine_step_parity.py (hunk 12 above)
Comment (tianhaox) [reply aic-core/rust/aiconfigurator-core/parity_tests/test_engine_step_parity.py]: Reworded in 93dec8ad3 — dropped the "audit finding #N" phrasing from both docstrings; the FFI contracts are now described on their own terms.

File: src/aiconfigurator/sdk/engine.py (hunk 13 above)
Comment (tianhaox) [reply src/aiconfigurator/sdk/engine.py]: Done in 93dec8ad3 — the comment is now a bulleted history mapping each bump to its source: 2 = v0.10.0 op-payload layout change, 3 = PR #1405 MTP-acceptance removal, 4 = PR #1355 MSA variant insertion merged on top of #1405 (each had independently claimed 3, so the merge took a fresh number). Note the constant moved to `aic-core/src/aiconfigurator_core/sdk/engine.py` with the #1322 relocation; `config.rs` carries the same history.
